core/src/main/java/org/elasticsearch/client/transport/TransportClientNodesService.java

```diff
@@ -470,78 +460,87 @@ protected void doSample() {
+                                        "failed to get local cluster state info for {}, disconnecting...", nodeToPing), e);
+                            }
                         }
-                    }
-                });
-            }
 
-            try {
+                        @Override
+                        protected void doRun() throws Exception {
+                            Transport.Connection pingConnection = null;
+                            if (nodes.contains(nodeToPing)) {
+                                try {
+                                    pingConnection = transportService.getConnection(nodeToPing);
+                                } catch (NodeNotConnectedException e) {
+                                    // will use a temp connection
+                                }
+                            }
+                            if (pingConnection == null) {
+                                logger.trace("connecting to cluster node [{}]", nodeToPing);
+                                connectionToClose = transportService.openConnection(nodeToPing, LISTED_NODES_PROFILE);
+                                pingConnection = connectionToClose;
+                            }
+                            transportService.sendRequest(pingConnection, ClusterStateAction.NAME,
+                                Requests.clusterStateRequest().clear().nodes(true).local(true),
+                                TransportRequestOptions.builder().withType(TransportRequestOptions.Type.STATE)
+                                    .withTimeout(pingTimeout).build(),
+                                new TransportResponseHandler<ClusterStateResponse>() {
+
+                                    @Override
+                                    public ClusterStateResponse newInstance() {
+                                        return new ClusterStateResponse();
+                                    }
+
+                                    @Override
+                                    public String executor() {
+                                        return ThreadPool.Names.SAME;
+                                    }
+
+                                    @Override
+                                    public void handleResponse(ClusterStateResponse response) {
+                                        clusterStateResponses.put(nodeToPing, response);
+                                        latch.countDown();
+                                    }
+
+                                    @Override
+                                    public void handleException(TransportException e) {
+                                        logger.info(
+                                            (Supplier<?>) () -> new ParameterizedMessage(
+                                                "failed to get local cluster state for {}, disconnecting...", nodeToPing), e);
+                                        latch.countDown();
```

Comment:
I am not sure but should we count down the latch once we notified the listener? I think it would be the better semantic.... maybe in a finally block?